src/jit/codegenarm64.cpp

```diff
@@ -2665,62 +2665,203 @@ void CodeGen::genJumpTable(GenTree* treeNode)
 // GT_LOCKADD, GT_XCHG, GT_XADD
 void CodeGen::genLockedInstructions(GenTreeOp* treeNode)
 {
-#if 0
-    GenTree* data       = treeNode->gtOp.gtOp2;
-    GenTree* addr       = treeNode->gtOp.gtOp1;
+    GenTree*  data      = treeNode->gtOp.gtOp2;
+    GenTree*  addr      = treeNode->gtOp.gtOp1;
     regNumber targetReg = treeNode->gtRegNum;
     regNumber dataReg   = data->gtRegNum;
     regNumber addrReg   = addr->gtRegNum;
-    instruction ins;
 
-    // all of these nodes implicitly do an indirection on op1
-    // so create a temporary node to feed into the pattern matching
-    GenTreeIndir i = indirForm(data->TypeGet(), addr);
-    genConsumeReg(addr);
+    regNumber exResultReg  = treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber storeDataReg = (treeNode->OperGet() == GT_XCHG) ? dataReg : treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber loadReg      = (targetReg != REG_NA) ? targetReg : storeDataReg;
 
     // The register allocator should have extended the lifetime of the address
     // so that it is not used as the target.
```

Comment:
Instead of addressing all of these requirements separately, it would be clearer, I think, to say:
```
// The register allocator should have extended the lifetimes of all input and internal registers so that
// none interfere with the target.